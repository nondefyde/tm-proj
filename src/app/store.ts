import { Action, configureStore, ThunkAction } from '@reduxjs/toolkit';
import personalityTestReducer from '../features/personality-test/PersonalityTestSlice';

export const store = configureStore({
  reducer: {
    personalityTest: personalityTestReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
