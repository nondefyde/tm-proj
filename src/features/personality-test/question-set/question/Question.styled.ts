import styled from 'styled-components';

export const QuestionStyledContainer = styled.div`
	padding: 10px;
`;

export const QuestionStyledTitle = styled.div`
	font-size: 20px;
	text-align: justify;
	text-weight: 600;
	margin-bottom: 30px;
`;


export const QuestionStyledButton = styled.button<{$isDefault?: boolean}>`
  margin-top: 30px;
  width: 100%;
  height: 50px;
  cursor: pointer;
  color: #ffffff;
  border: 1px solid #cccccc;
  
  ${(props) => props.$isDefault ?
    `background-color: #f6f6f6;
     color: #000000;
    `
    : `background-color: #226bef;
       color: #ffffff;
    `}
  :disabled {
    opacity: 0.5;
    cursor: no-drop;
  }
`;

export const StyledButtons = styled.div`
  display: flex;
  flex-direction: row;
  gap: 10px;
  margin-bottom: 20px;
`;