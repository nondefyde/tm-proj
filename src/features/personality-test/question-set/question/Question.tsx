import React, { useState } from 'react';
import { AnswerType, QuestionType, QuestionTypeItem } from './index';
import { QuestionStyledButton, QuestionStyledContainer, QuestionStyledTitle, StyledButtons } from './Question.styled';
import { QuestionItem } from '../question-item';

export type QuestionProps = {
  currentIndex: number,
  answer?: AnswerType,
  question: QuestionType,
  onPrevious: (answer?: AnswerType) => void
  onNext: (answer?: AnswerType) => void
  answerQuestion: (answer: AnswerType) => void
}

export function Question(props: QuestionProps) {
  const { question, answer, answerQuestion, onPrevious, onNext, currentIndex } = props;
  
  const [selected, setSelected] = useState<AnswerType | undefined>(answer);
  
  console.log('selected ::: ', selected);
  
  const onSelected = (answer: AnswerType) => {
    setSelected(answer);
    answerQuestion(answer)
  }
  
  return (
    <QuestionStyledContainer>
      <QuestionStyledTitle>
        {question.title}
      </QuestionStyledTitle>
      {question.items.map((item: QuestionTypeItem) => {
        return <QuestionItem
          selectedAnswer={selected}
          questionId={question.id}
          onItemClicked={onSelected}
          key={item.id} item={item} />
      })}
      <StyledButtons>
        {currentIndex > 0 &&
          <QuestionStyledButton
            data-testid={'previous-button'}
            disabled={currentIndex <= 0}
            $isDefault={true}
            onClick={() => onPrevious(selected)}>
            Previous Question
          </QuestionStyledButton>
        }
        <QuestionStyledButton
          data-testid={'next-button'}
          disabled={!selected}
          onClick={() => onNext(selected)}>
          Next Question
        </QuestionStyledButton>
      </StyledButtons>
    </QuestionStyledContainer>
  );
}
