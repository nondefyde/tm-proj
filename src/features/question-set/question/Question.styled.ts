import styled from 'styled-components';

export const QuestionStyledContainer = styled.div`
	padding: 10px;
`;

export const QuestionStyledTitle = styled.div`
	font-size: 25px;
	text-align: justify;
	text-weight: 600;
	font-family: "Times New Roman", Times, serif;
	margin-bottom: 30px;
`;


export const QuestionStyledButton = styled.button`
	margin-top: 30px;
	width: 100%;
	height: 50px
`;