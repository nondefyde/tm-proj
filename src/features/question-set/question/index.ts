export * from './Question'

export const enum QuestionEnumType {
	MULTIPLE = 'multipleSelection',
	SINGLE = 'singleSelection'
}

export type QuestionTypeItem = {
	id: string,
	identifier: string,
	description: string,
}

export type QuestionType = {
	id: string,
	title: string,
	type: QuestionEnumType,
	items: QuestionTypeItem[]
}
