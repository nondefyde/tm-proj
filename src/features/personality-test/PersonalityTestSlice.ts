import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchQuestion } from './question-set/questionsAPI';
import { AnswerType, QuestionType } from './question-set/question';
import { RootState } from '../../app/store';

export interface QuestionSetState {
  questions: QuestionType[];
  current?: QuestionType;
  currentIndex: number;
  answers: Record<string, AnswerType>;
  status: 'idle' | 'loading' | 'failed';
}

const initialState: QuestionSetState = {
  questions: [],
  currentIndex: 0,
  current: undefined,
  answers: {},
  status: 'idle',
};

export const fetchQuestionAsync = createAsyncThunk(
  'questions/fetchQuestions',
  async () => {
    const response = await fetchQuestion();
    // The value we return becomes the `fulfilled` action payload
    return response.data;
  }
);

export const personalityTestSlice = createSlice({
  name: 'personalityTest',
  initialState,
  // The `reducers` field lets us define reducers and generate associated actions
  reducers: {
    // Use the PayloadAction type to declare the contents of `action.payload`
    selectNextQuestion: (state, action: PayloadAction<string>) => {
      const index = state.currentIndex + 1
      if (state.questions[index]) {
        state.current = state.questions[index];
        state.currentIndex = index;
      }
    },
    selectPreviousQuestion: (state) => {
      const index = state.currentIndex - 1
      if(state.questions[index]) {
        state.current = state.questions[index];
        state.currentIndex = index;
      }
    },
    answerQuestion: (state, action: PayloadAction<AnswerType>) => {
      state.answers[action.payload.questionId] = action.payload;
    },
  
    resetTest: (state) => {
      state.current = undefined;
      state.currentIndex = 0;
      state.answers = {};
    },
  },
  // The `extraReducers` field lets the slice handle actions defined elsewhere,
  // including actions generated by createAsyncThunk or in other slices.
  extraReducers: (builder) => {
    builder
      .addCase(fetchQuestionAsync.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchQuestionAsync.fulfilled, (state, action) => {
        state.status = 'idle';
        state.questions = action.payload;
        if(action.payload.length) {
          state.current = state.questions[0];
          state.currentIndex = 0;
        }
      })
      .addCase(fetchQuestionAsync.rejected, (state) => {
        state.status = 'failed';
      });
  },
});

export const { answerQuestion, resetTest, selectNextQuestion, selectPreviousQuestion } = personalityTestSlice.actions;

export default personalityTestSlice.reducer;
