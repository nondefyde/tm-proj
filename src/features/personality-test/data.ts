import {QuestionEnumType, QuestionType, ResultType} from './question-set/question';


export const allQuestions: QuestionType[] = [
  {
    id: '1',
    type: QuestionEnumType.MULTIPLE,
    title: 'You’re really busy at work and a colleague is telling you their life story and personal woes. You:',
    items: [
      {
        id: '1',
        identifier: 'A',
        description: 'Don’t dare to interrupt them'
      },
      {
        id: '2',
        identifier: 'B',
        description: 'Think it’s more important to give them some of your time; work can wait'
      },
      {
        id: '3',
        identifier: 'C',
        description: 'Listen, but with only with half an ear'
      },
      {
        id: '4',
        identifier: 'D',
        description: 'Interrupt and explain that you are really busy at the moment'
      }
    ]
  },
  {
    id: '2',
    type: QuestionEnumType.MULTIPLE,
    title: 'You’ve been sitting in the doctor’s waiting room for more than 25 minutes. You:',
    items: [
      {
        id: '1',
        identifier: 'A',
        description: 'Look at your watch every two minutes'
      },
      {
        id: '2',
        identifier: 'B',
        description: 'Bubble with inner anger, but keep quiet'
      },
      {
        id: '3',
        identifier: 'C',
        description: 'Explain to other equally impatient people in the room that the doctor is always running late'
      },
      {
        id: '4',
        identifier: 'D',
        description: 'Complain in a loud voice, while tapping your foot impatiently'
      }
    ]
  },
  {
    id: '3',
    type: QuestionEnumType.MULTIPLE,
    title: 'You’re having an animated discussion with a colleague regarding a project that you’re in charge of. You:\n',
    items: [
      {
        id: '1',
        identifier: 'A',
        description: 'Don’t dare contradict them'
      },
      {
        id: '2',
        identifier: 'B',
        description: 'Think that they are obviously right'
      },
      {
        id: '3',
        identifier: 'C',
        description: 'Defend your own point of view, tooth and nail'
      },
      {
        id: '4',
        identifier: 'D',
        description: 'Continuously interrupt your colleague'
      }
    ]
  }
]

export const resultData: ResultType[] = [
  {
    type: 'introvert',
    title: 'Your Result\n' +
        'You are more of an introvert',
    description: 'You feel that living alone is to live happily, and you prefer hiding in a crowd rather than standing' +
        ' out in one. You are perpetually tormented by the idea of doing things wrong, not understanding or not being ' +
        'alert enough or intelligent enough to do what others expect of you. You lack in self-confidence and you seem ' +
        ' So, you try to remain the most transparent and discreet possible. How far will you take this logic of' +
        ' self-denigration? Try and stop projecting onto others the bad image you have about yourself. ' +
        'Have a good look around you and you’ll see that they too have faults, weaknesses and shortcomings,' +
      'Have a good look around you and you’ll see that they too have faults, weaknesses and shortcomings,' +
        ' so stop finding excuses for everything. Maybe you were brought up in an atmosphere of ‘You’ll never make' +
      ' the grade’ as a child — a poison that you need progressively to get out of your system. Learn to look after' +
      ' your own interests — everyone else does, so why not you? You too have desires, dreams and opinions — express' +
      ' them. If you think that by saying no or thinking differently from others that you will no longer be loved,' +
      ' it just isn’t so. Others will learn to respect you because you respect yourself now.'
  },
  {
    type: 'introvert',
    title: 'Your Result\n' +
        'You are more of an Extrovert',
    description: 'You feel that living alone is to live happily, and you prefer hiding in a crowd rather than standing' +
        ' so stop finding excuses for everything. Maybe you were brought up in an atmosphere of ‘You’ll never make the' +
        ' grade’ as a child — a poison that you need progressively to get out of your system. Learn to look after your' +
      ' out in one. You are perpetually tormented by the idea of doing things wrong, not understanding or not being ' +
      'alert enough or intelligent enough to do what others expect of you. You lack in self-confidence and you seem ' +
      'Have a good look around you and you’ll see that they too have faults, weaknesses and shortcomings,' +
        ' own interests — everyone else does, so why not you? You too have desires, dreams and opinions — express them.' +
        ' If you think that by saying no or thinking differently from others that you will no longer be loved, it just' +
        ' isn’t so. Others will learn to respect you because you respect yourself'
  }
]