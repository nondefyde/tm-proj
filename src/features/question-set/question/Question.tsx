import React  from 'react';
import { QuestionType, QuestionTypeItem } from './index';
import { QuestionStyledButton, QuestionStyledContainer, QuestionStyledTitle } from './Question.styled';
import { QuestionItem } from '../question-item';

export type QuestionProps = {
  question: QuestionType,
}

export function Question(props: QuestionProps) {
  const {question} = props;
  
  const onNextButtonClicked = () => {}
  
  const onItemClicked = (item: QuestionTypeItem) => {
    console.log('OUTPUT item ::: ', item);
  }
  
  return (
    <QuestionStyledContainer>
      <QuestionStyledTitle>
        {question.title}
      </QuestionStyledTitle>
      {question.items.map((item: QuestionTypeItem) => {
        return <QuestionItem
          onItemClicked={onItemClicked}
          key={item.id} item={item} />
      })}
      <QuestionStyledButton onClick={onNextButtonClicked}>
        Next Question
      </QuestionStyledButton>
    </QuestionStyledContainer>
  );
}
