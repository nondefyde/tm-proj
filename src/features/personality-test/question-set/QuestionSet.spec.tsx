import { render } from '@testing-library/react'
import { QuestionSet } from './index';
import * as hooks from '../../../app/hooks';
import { allQuestions } from '../data';
import * as QuestionObject from './question';
import { answerQuestion, selectNextQuestion, selectPreviousQuestion } from '../PersonalityTestSlice';

const mockDispatch = jest.fn();
jest.mock('../../../app/hooks', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: () => mockDispatch
}));
jest.mock('./question')

describe('Question Component', () => {
  const useAppDispatchMock = jest.spyOn(hooks, 'useAppDispatch');
  const useAppSelectorMock = jest.spyOn(hooks, 'useAppSelector');
  
  const initialData = {
    personalityTest: {
      result: {},
      answers: {},
      currentIndex: 0,
      status: 'idle',
      questions: allQuestions,
      current: {
        '1': true
      }
    },
  }
  
  let onPrevious: any;
  let onNext: any;
  let onAnswerQuestion: any;
  
  let answerMock = { questionId: '1', item: { id: '1', identifier: 'A', description: '' } };
  
  afterEach(() => {
    useAppSelectorMock.mockClear();
    useAppDispatchMock.mockClear();
  });
  
  beforeEach(() => {
    const QuestionMock = jest.spyOn(QuestionObject, 'Question')
    QuestionMock.mockImplementation((props) => {
      onPrevious = props?.onPrevious;
      onNext = props?.onNext;
      onAnswerQuestion = props?.answerQuestion;
      return <>***Question***</>
    })
    
    useAppDispatchMock.mockReturnValue(mockDispatch);
    useAppSelectorMock.mockImplementation((selectorFn: any) => selectorFn(initialData));
  });
  
  it('should test snapshot', () => {
    const container = render(<QuestionSet/>)
    expect(container).toMatchSnapshot();
  });
  
  it('should call selectNextQuestion', () => {
    render(<QuestionSet/>);
    onNext(answerMock);
    expect(mockDispatch).toHaveBeenCalledWith(selectNextQuestion('1'));
  });
  
  it('should call selectPreviousQuestion', () => {
    render(<QuestionSet/>);
    onPrevious();
    expect(mockDispatch).toHaveBeenCalledWith(selectPreviousQuestion());
  });
  
  it('should call answerQuestion', () => {
    render(<QuestionSet/>);
    onAnswerQuestion(answerMock);
    expect(mockDispatch).toHaveBeenCalledWith(answerQuestion(answerMock));
  });
  
  it('should show loading text', () => {
    useAppSelectorMock.mockImplementation((selectorFn: any) => selectorFn({
      personalityTest: {
        ...initialData.personalityTest,
        status: 'loading',
        questions: [],
      }
    }));
    const { getByText } = render(<QuestionSet/>)
    const NoQuestionText = getByText('Loading...')
    expect(NoQuestionText).toBeInTheDocument();
  });
  
  it('should show questions', () => {
    useAppSelectorMock.mockImplementation((selectorFn: any) => selectorFn({
      personalityTest: {
        ...initialData.personalityTest,
        current: undefined,
        status: 'idle',
      }
    }));
    const { getByText } = render(<QuestionSet/>)
    const NoQuestionText = getByText('No Question')
    expect(NoQuestionText).toBeInTheDocument();
  });
});
