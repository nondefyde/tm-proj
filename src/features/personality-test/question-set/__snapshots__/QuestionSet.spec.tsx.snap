// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question Component should test snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bcXHqe ldaQpu"
      >
        <div
          class="sc-gswNZR eIsmLB"
        >
          Question 1/14
        </div>
        <div
          class="sc-dkrFOg iXsOjN"
        >
          <div
            class="sc-hLBbgP iuyxgk"
          >
            You’re really busy at work and a colleague is telling you their life story and personal woes. You:
          </div>
          <div
            class="sc-jSUZER gbescb"
          >
            <div
              class="sc-gKPRtg gtuoho"
            >
              <span>
                A
              </span>
            </div>
            <div
              class="sc-iBYQkv kiwGZE"
            >
              Don’t dare to interrupt them
            </div>
          </div>
          <div
            class="sc-jSUZER gbescb"
          >
            <div
              class="sc-gKPRtg gtuoho"
            >
              <span>
                B
              </span>
            </div>
            <div
              class="sc-iBYQkv kiwGZE"
            >
              Think it’s more important to give them some of your time; work can wait
            </div>
          </div>
          <div
            class="sc-jSUZER gbescb"
          >
            <div
              class="sc-gKPRtg gtuoho"
            >
              <span>
                C
              </span>
            </div>
            <div
              class="sc-iBYQkv kiwGZE"
            >
              Listen, but with only with half an ear
            </div>
          </div>
          <div
            class="sc-jSUZER gbescb"
          >
            <div
              class="sc-gKPRtg gtuoho"
            >
              <span>
                D
              </span>
            </div>
            <div
              class="sc-iBYQkv kiwGZE"
            >
              Interrupt and explain that you are really busy at the moment
            </div>
          </div>
          <button
            class="sc-eDvSVe fWjGBZ"
          >
            Next Question
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bcXHqe ldaQpu"
    >
      <div
        class="sc-gswNZR eIsmLB"
      >
        Question 1/14
      </div>
      <div
        class="sc-dkrFOg iXsOjN"
      >
        <div
          class="sc-hLBbgP iuyxgk"
        >
          You’re really busy at work and a colleague is telling you their life story and personal woes. You:
        </div>
        <div
          class="sc-jSUZER gbescb"
        >
          <div
            class="sc-gKPRtg gtuoho"
          >
            <span>
              A
            </span>
          </div>
          <div
            class="sc-iBYQkv kiwGZE"
          >
            Don’t dare to interrupt them
          </div>
        </div>
        <div
          class="sc-jSUZER gbescb"
        >
          <div
            class="sc-gKPRtg gtuoho"
          >
            <span>
              B
            </span>
          </div>
          <div
            class="sc-iBYQkv kiwGZE"
          >
            Think it’s more important to give them some of your time; work can wait
          </div>
        </div>
        <div
          class="sc-jSUZER gbescb"
        >
          <div
            class="sc-gKPRtg gtuoho"
          >
            <span>
              C
            </span>
          </div>
          <div
            class="sc-iBYQkv kiwGZE"
          >
            Listen, but with only with half an ear
          </div>
        </div>
        <div
          class="sc-jSUZER gbescb"
        >
          <div
            class="sc-gKPRtg gtuoho"
          >
            <span>
              D
            </span>
          </div>
          <div
            class="sc-iBYQkv kiwGZE"
          >
            Interrupt and explain that you are really busy at the moment
          </div>
        </div>
        <button
          class="sc-eDvSVe fWjGBZ"
        >
          Next Question
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
